docs/designers-developers/developers/data/data-core-block-editor.md

```diff
@@ -1172,6 +1172,7 @@ _Parameters_
 
 -   _rootClientId_ `?string`: Optional root client ID of block list on which to insert.
 -   _index_ `?number`: Index at which block should be inserted.
+-   _canReplaceBlock_ `?boolean`: can the insertion point replace the block at index.
```

Comment:
It is already derived from `Inserter`’s local state via [getShouldReplaceBlock](https://github.com/WordPress/gutenberg/pull/18791/files#diff-8aa4726165b0e0cdb260abe61e3329ceR97) function. But we need to pass this state to the `BlockList` so it [hides the blocks](https://github.com/WordPress/gutenberg/pull/18791/files#diff-260b0f5b413f5ac735a6935b2f8acc34R170) where the `BlockInsertionPoint` will replace. `Inserter` and `BlockList` don't share a very close parent so instead of lifting the state all the way up to their common parent and drilling down, I chose to store this state it in the store :) It might be used in web for a similar behavior with "Add Block Before/After" for where a block in the list could be hidden when it will be replaced and `BlockInsertionPoint` is shown in it's place.

But I agree this state is now not about only showing/hiding `BlockInsertionPoint` but it's even about showing/hiding a block in the list.  

But maybe there is another way without changing the web store. Instead of hiding a block we can try to remove the block completely from `BlockList` when `BlockInsertionPoint` is shown and the block is going to be replaced. If the block insertion is cancelled (via closing the menu) we can re-insert the block again. It should be easy to re-insert the block again because the replace only happens for unmodified default blocks (empty paragraph) right now. If a new block is picked from the menu instead of cancelling we can just call [insert](https://github.com/WordPress/gutenberg/pull/18791/files#diff-e7138e42f97fc785033583bf87f62519R166) instead of [replace](https://github.com/WordPress/gutenberg/pull/18791/files#diff-e7138e42f97fc785033583bf87f62519R164) to add the new block. This way we might even remove [show/hide block](https://github.com/WordPress/gutenberg/pull/18791/files#diff-260b0f5b413f5ac735a6935b2f8acc34R262) logic completely from `BlockList`.  

What do you think?

Reply:
> What do you think?

I think it sounds quite hacky, but that's actually a clever way to do it without compromising on the functionalities, I like it 👍

I say it's hacky because this approach won't work if `isSelectedBlockReplaceable` becomes more complex than just `isSelectedUnmodifiedDefaultBlock`. Maybe we will then be able to "store" the type of block that was removed in the Inserter state, but then it feels even more hacky 🤔

I think it's fine for now to have more custom code in the native Inserter since our version does more than the web. Adding new attributes to the store that are not presently needed is hardly justifiable at this point in the project.

Reply:
I just pushed the mentioned changes (and also reverted scrollview related changes), feel free to check it out again :)